tests/30rooms/14forget.pl

```diff
@@ -0,0 +1,278 @@
+            params => {
+               limit => 1,
+               dir   => 'b'
+            },
+      )})->then( sub {
+         my ( $body ) = @_;
+
+         require_json_keys( $body, qw( chunk ) );
+         log_if_fail "Chunk", $body->{chunk};
+         $body->{chunk}[0]{content}{body} eq "sup" or die "Wrong message";
+
+         Future->done(1);
+      })->then( sub {
+         matrix_forget_room( $user, $room_id )
+      })->then( sub {
+         matrix_get_room_state( $creator, $room_id,
+            type => "m.room.member",
+            state_key => $user->user_id
+         )
+      })->then( sub {
+         my ( $event ) = @_;
+
+         $event->{membership} eq "leave" or die "Wrong membership state";
+
+         do_request_json_for( $user,
+            method => "GET",
+            uri    => "/api/v1/rooms/$room_id/messages",
+            params => {
+               limit => 1,
+               dir   => 'b'
+            },
+         )->main::expect_http_403;
+      })
+   };
+
+test "Can re-join room if re-invited - history_visibility = shared",
+   requires => [ local_user_fixture(), local_user_fixture() ],
+
+   do => sub {
+      my ( $creator, $user ) = @_;
+
+      my ( $room_id );
+
+      matrix_create_room( $creator, invite => [ $user->user_id ] )->then( sub {
+         ( $room_id ) = @_;
+
+         log_if_fail "room_id", $room_id;
+
+         matrix_put_room_state( $creator, $room_id,
+            type => "m.room.join_rules",
+            state_key => "",
+            content => {
+               join_rule => "invite",
+            }
+         )
+      })->then( sub {
+         matrix_join_room( $user, $room_id );
+      })->then( sub {
+         matrix_send_room_text_message( $creator, $room_id, body => "before leave" );
+      })->then( sub {
+         do_request_json_for( $user,
+            method => "GET",
+            uri    => "/api/v1/rooms/$room_id/messages",
+            params => {
+               limit => 100,
+               dir   => 'b'
+            },
+         );
+      })->then( sub {
+         matrix_forget_room( $user, $room_id );
+      })->then( sub {
+         matrix_join_room( $user, $room_id )->main::expect_http_403;
+      })->then( sub {
+         matrix_invite_user_to_room( $creator, $user, $room_id );
+      })->then( sub {
+         matrix_join_room( $user, $room_id );
+      })->then( sub {
+         do_request_json_for( $user,
+            method => "GET",
+            uri    => "/api/v1/rooms/$room_id/messages",
+            params => {
+               limit => 100,
+               dir   => 'b'
+            },
+         );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         my $seen = 0;
+
+         foreach my $event (@{ $body->{chunk} }) {
+            $seen = 1 if $event->{type} eq "m.room.message" && $event->{content}->{body} eq "before leave";
+         }
+
+         die "Should have seen before leave message" unless $seen;
```

Comment:
This might be a use for `List::Util::first`:

    first {
       $_->{type} eq "m.room.message" and $_->{content}{body} eq "before leave"
    } @{ $body->{chunk} } or
       die "Should have seen before leave message";